src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs

```diff
@@ -641,47 +641,51 @@ internal NamedTypeSymbol LookupTopLevelMetadataType(ref MetadataTypeName emitted
-        internal AssemblySymbol GetAssemblyForForwardedType(ref MetadataTypeName fullName)
+        internal (AssemblySymbol FirstSymbol, AssemblySymbol SecondSymbol) GetAssembliesForForwardedType(ref MetadataTypeName fullName)
         {
             string matchedName;
-            AssemblyReferenceHandle assemblyRef = Module.GetAssemblyForForwardedType(fullName.FullName, ignoreCase: false, matchedName: out matchedName);
-            return assemblyRef.IsNil ? null : this.GetReferencedAssemblySymbol(assemblyRef);
+            (int firstIndex, int secondIndex) = this.Module.GetAssemblyRefsForForwardedType(fullName.FullName, ignoreCase: false, matchedName: out matchedName);
+
+            var firstSymbol = firstIndex < 0 ? null : GetReferencedAssemblySymbol(firstIndex);
+            var secondSymbol = secondIndex < 0 ? null : GetReferencedAssemblySymbol(secondIndex);
+
+            return (firstSymbol, secondSymbol);
         }
 
         internal IEnumerable<NamedTypeSymbol> GetForwardedTypes()
         {
-            foreach (KeyValuePair<string, AssemblyReferenceHandle> forwarder in Module.GetForwardedTypes())
+            foreach (KeyValuePair<string, (int FirstIndex, int SecondIndex)> forwarder in Module.GetForwardedTypes())
             {
-                var assemblySymbol = this.GetReferencedAssemblySymbol(forwarder.Value);
-                if ((object)assemblySymbol == null)
+                if (forwarder.Value.FirstIndex < 0)
                 {
                     continue;
                 }
+
                 var name = MetadataTypeName.FromFullName(forwarder.Key);
-                yield return assemblySymbol.LookupTopLevelMetadataType(ref name, digThroughForwardedTypes: true);
-            }
-        }
 
-        private AssemblySymbol GetReferencedAssemblySymbol(AssemblyReferenceHandle assemblyRef)
-        {
-            int referencedAssemblyIndex;
-            try
-            {
-                referencedAssemblyIndex = Module.GetAssemblyReferenceIndexOrThrow(assemblyRef);
-            }
-            catch (BadImageFormatException)
-            {
-                return null;
+                AssemblySymbol firstSymbol = this.GetReferencedAssemblySymbol(forwarder.Value.FirstIndex);
+                Debug.Assert((object)firstSymbol != null, "Invalid indexes (out of bound) are discarded during reading metadata in PEModule.EnsureForwardTypeToAssemblyMap()");
+
+                if (forwarder.Value.SecondIndex >= 0)
+                {
+                    var secondSymbol = this.GetReferencedAssemblySymbol(forwarder.Value.SecondIndex);
+                    Debug.Assert((object)secondSymbol != null, "Invalid indexes (out of bound) are discarded during reading metadata in PEModule.EnsureForwardTypeToAssemblyMap()");
+
+                    var forwardingErrorInfo = new DiagnosticInfo(MessageProvider.Instance, (int)ErrorCode.ERR_TypeForwardedToMultipleAssemblies, forwarder.Key, firstSymbol.Name, secondSymbol.Name);
```

Comment:
>var forwardingErrorInfo = new DiagnosticInfo(MessageProvider.Instance, (int)ErrorCode.ERR_TypeForwardedToMultipleAssemblies, forwarder.Key, firstSymbol.Name, secondSymbol.Name); [](start = 20, length = 177)

It looks like we are duplicating code with PEAssemblySymbol, consider adding a helper similar to the one we have for a cycle and reusing it.